File: Week 02/id_063/Java/LeetCode_589_063.java
import java.util.LinkedList;
import java.util.List;

/*
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

/*
思路
简单递归前序遍历
*/

public class Solution {
    private List<Integer> dfs(List<Integer> result, Node curNode) {
        if (curNode == null) {
            return result;
        }

        result.add(curNode.val);
        if (curNode.children != null) {
            for (Node subNode : curNode.children) {
                dfs(result, subNode);
            }
        }

        return result;
    }

    public List<Integer> preorder(Node root) {
        return dfs(new LinkedList<Integer>(), root);
    }
}

File: Week 02/id_063/C++/LeetCode_589_063.cpp
#include <vector>
using namespace std;

/*
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:

    void dfs(vector<int>& result, Node* p_cur_node) {
        if (p_cur_node == nullptr) {
            return;
        }

        result.push_back(p_cur_node->val);
        for (Node* p_sub_node : p_cur_node->children) {
            dfs(result, p_sub_node);
        }
    }

    vector<int> preorder(Node* root) {
        vector<int> v;
        dfs(v, root);

        return v;
    }
};